publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java

```diff
@@ -400,33 +420,51 @@ private boolean hasExchange() throws RemRemPublishException {
-                        System.exit(-3);
+    public void send(String routingKey, String msg) throws  IOException ,NackException ,TimeoutException, ChannelClosureException{
+            Channel channel = giveMeRandomChannel();
+            channel.addShutdownListener(new ShutdownListener() {
+                public void shutdownCompleted(ShutdownSignalException cause) {
+                    // Beware that proper synchronization is needed here
+                    if (cause.isInitiatedByApplication()) {
+                        log.debug("Shutdown is initiated by application. Ignoring it.");
+                    } else {
+                        log.error("Shutdown is NOT initiated by application.");
+                        log.error(cause.getMessage());
+                        boolean cliMode = Boolean.getBoolean(PropertiesConfig.CLI_MODE);
+                        if (cliMode) {
+                            System.exit(-3);
+                        }
                     }
                 }
+            });
+            BasicProperties msgProps = MessageProperties.BASIC;
+            if (usePersitance)
+                msgProps = MessageProperties.PERSISTENT_BASIC;
+        try {
+            channel.basicPublish(exchangeName, routingKey, msgProps, msg.getBytes());
+            log.info("Published message with size {} bytes on exchange '{}' with routing key '{}'",
+                    msg.getBytes().length, exchangeName, routingKey);
+            if (WaitForConfirmstimeOut == null || WaitForConfirmstimeOut == 0) {
+                WaitForConfirmstimeOut = 5000L;
             }
-        });
-
-        BasicProperties msgProps = MessageProperties.BASIC;
-        if (usePersitance)
-            msgProps = MessageProperties.PERSISTENT_BASIC;
-
-        channel.basicPublish(exchangeName, routingKey, msgProps, msg.getBytes());
-        log.info("Published message with size {} bytes on exchange '{}' with routing key '{}'", msg.getBytes().length,
-                exchangeName, routingKey);
+            channel.waitForConfirmsOrDie(WaitForConfirmstimeOut);
+            log.info("channel seq number is.............."+channel.getNextPublishSeqNo());
+
+        } catch (InterruptedException | IOException e) {
+            log.error("Failed to publish message due to" + e.getMessage());
+            throw new NackException("the message is nacked due to " + e.getMessage());
+        } catch (TimeoutException e) {
+            log.error("Failed to publish message due to" + e.getMessage());
+            throw new TimeoutException("Timeout waiting for ACK  " + e.getMessage());
+        } catch (Exception e) {
+            log.error("Failed to publish message due to " + e.getMessage());
+            if(!channel.isOpen()&& rabbitConnection.isOpen()){
+                throw new ChannelClosureException("Failed to publish message due to " + e.getMessage());
+            }
+            throw new IOException("Failed to publish message due to " + e.getMessage());
```

Comment:
The newly thrown exceptions should include the original exception as a root cause.